webapp/platform/client/src/websocket.test.ts

```diff
@@ -0,0 +1,206 @@
+
+    public onopen: (() => void) | null = null;
+    public onclose: (() => void) | null = null;
+    public onerror: (() => void) | null = null;
+    public onmessage: ((evt: any) => void) | null = null;
+
+    open() {
+        this.readyState = WebSocket.OPEN;
+        if (this.onopen) {
+            this.onopen();
+        }
+    }
+
+    close() {
+        this.readyState = WebSocket.CLOSED;
+        if (this.onclose) {
+            this.onclose();
+        }
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    send(msg: any) { }
+    addEventListener() { }
+    removeEventListener() { }
+    dispatchEvent(): boolean {
+        return false;
+    }
+}
+
+describe('websocketclient', () => {
+    test('should call callbacks', () => {
+        const mockWebSocket = new MockWebSocket();
+
+        const client = new WebSocketClient({
+            newWebSocketFn: (url: string) => {
+                mockWebSocket.url = url;
+                return mockWebSocket;
+            },
+        });
+        client.initialize('mock.url');
+
+        expect(mockWebSocket.onopen).toBeTruthy();
+        mockWebSocket.onopen = jest.fn();
+        expect(mockWebSocket.onclose).toBeTruthy();
+        mockWebSocket.onclose = jest.fn();
+
+        mockWebSocket.open();
+
+        expect(mockWebSocket.onopen).toHaveBeenCalled();
+        expect(mockWebSocket.readyState).toBe(mockWebSocket.OPEN);
+
+        mockWebSocket.close();
+
+        expect(mockWebSocket.onclose).toHaveBeenCalled();
+        expect(mockWebSocket.readyState).toBe(mockWebSocket.CLOSED);
+
+        client.close();
+    });
+
+    test('should reconnect on websocket close', (done) => {
+        const mockWebSocket = new MockWebSocket();
+        mockWebSocket.open = jest.fn(mockWebSocket.open);
+
+        const client = new WebSocketClient({
+            newWebSocketFn: (url: string) => {
+                mockWebSocket.url = url;
+                mockWebSocket.open();
+                return mockWebSocket;
+            },
+            minWebSocketRetryTime: 1,
+            reconnectJitterRange: 1,
+        });
+        client.initialize('mock.url');
+
+        mockWebSocket.close();
+
+        setTimeout(() => {
+            client.close();
+            expect(mockWebSocket.open).toHaveBeenCalledTimes(2);
+            done();
+        }, 10);
+    });
+
+    test('should close during reconnection delay', (done) => {
+        const mockWebSocket = new MockWebSocket();
+        mockWebSocket.open = jest.fn(mockWebSocket.open);
+
+        const client = new WebSocketClient({
+            newWebSocketFn: (url: string) => {
+                mockWebSocket.url = url;
+                if (mockWebSocket.onopen) {
+                    mockWebSocket.open();
+                }
+                return mockWebSocket;
+            },
+            minWebSocketRetryTime: 50,
+            reconnectJitterRange: 1,
+        });
+        client.initialize = jest.fn(client.initialize);
```

Comment:
Instead of replacing the method with the mock, you might be able to use `const spy = jest.spyOn(client, 'initialize')` to accomplish the same thing without overwriting anything on the client directly. I think you can then call `expect(spy).toHaveBeenCalled()` like you're doing now

Reply:
Good suggestion - done.